game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java

```diff
@@ -1248,65 +1248,57 @@ public void execute(final ExecutionStack stack, final IDelegateBridge bridge) {
    * @return the removed units
    */
   @Override
-  public Collection<Unit> removeNonCombatants(final Side side) {
+  public Collection<Unit> removeNonCombatants(
+      final Side side, final boolean removeAaUnitsThatFireThisRound) {
+    final List<Unit> combatants = calculateCombatants(side, removeAaUnitsThatFireThisRound);
+    final Collection<Unit> nonCombatants;
     if (side == DEFENSE) {
-      final List<Unit> notRemovedDefending =
-          removeNonCombatants(defendingUnits, attackingUnits, false, true);
-      final Collection<Unit> toRemoveDefending =
-          CollectionUtils.difference(defendingUnits, notRemovedDefending);
-      defendingUnits = notRemovedDefending;
-      return toRemoveDefending;
+      nonCombatants = CollectionUtils.difference(defendingUnits, combatants);
+      defendingUnits = combatants;
     } else {
-      final List<Unit> notRemovedAttacking =
-          removeNonCombatants(attackingUnits, defendingUnits, true, true);
-      final Collection<Unit> toRemoveAttacking =
-          CollectionUtils.difference(attackingUnits, notRemovedAttacking);
-      attackingUnits = notRemovedAttacking;
-      return toRemoveAttacking;
+      nonCombatants = CollectionUtils.difference(attackingUnits, combatants);
+      attackingUnits = combatants;
     }
+    return nonCombatants;
   }
 
   /**
-   * Returns only the relevant non-combatant units present in the specified collection.
+   * Returns only relevant combatant units present in the specified collection.
    *
-   * @return a collection containing all the combatants in units non-combatants include such things
-   *     as factories, aa guns, land units in a water battle.
+   * <p>Non-combatants include such things as:
+   *
+   * <ul>
+   *   <li>factories
+   *   <li>aa guns that won't fire any more
+   *   <li>land units in a water battle
+   *   <li>sea units in a land battle
+   *   <li>units that will be captured on entering
+   * </ul>
    */
-  private List<Unit> removeNonCombatants(
-      final Collection<Unit> units,
-      final Collection<Unit> enemyUnits,
-      final boolean attacking,
-      final boolean removeForNextRound) {
-    final List<Unit> unitList = new ArrayList<>(units);
-    if (battleSite.isWater()) {
-      unitList.removeAll(CollectionUtils.getMatches(unitList, Matches.unitIsLand()));
```

Comment:
This match is already being done in `Matches.unitCanBeInBattle` so it isn't needed here.